File: Android/app/src/main/java/com/rpinferetti/picar/TCPSocket.java
package com.rpinferetti.picar;

import android.os.AsyncTask;
import android.util.Log;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.net.InetAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

public class TCPSocket {
    private static final String TAG = "TCPSocket";

    private OnConnectSocketListener mListener;

    private static Socket mSocket;
    private static Writer mBufferOut;


    void connect(String address, int port) {
        ConnectTask task = new ConnectTask(address, port);
        task.setOnConnectTaskListener(new ConnectTask.OnConnectTaskListener() {
            @Override
            public void onSuccess() {
                mListener.onSuccess();
            }

            @Override
            public void onFailure() {
                mListener.onFailure();
            }
        });
        task.execute();
    }

    public void disconnect() {
        try {
            mBufferOut.close();
            mSocket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void sendMessage(final String msg) {
        if (mSocket != null) {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    if (mBufferOut != null) {
                        try {
                            Log.d(TAG, msg);
                            mBufferOut.write(msg);
                            mBufferOut.flush();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }).start();
        }
    }

    void setOnConnectSocketListener(OnConnectSocketListener listener) {
        mListener = listener;
    }

    public interface OnConnectSocketListener {
        void onSuccess();

        void onFailure();
    }

    public static class ConnectTask extends AsyncTask<Void, Void, Void> {
        private String address;
        private int port;
        private OnConnectTaskListener mListener;

        ConnectTask(String address, int port) {
            this.address = address;
            this.port = port;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            try {
                InetAddress serverAddr = InetAddress.getByName(address);
                mSocket = new Socket(serverAddr, port);
                mBufferOut = new BufferedWriter(new OutputStreamWriter(mSocket.getOutputStream(), StandardCharsets.UTF_8));
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            if (mSocket != null) {
                mListener.onSuccess();
            } else {
                mListener.onFailure();
            }
        }

        void setOnConnectTaskListener(OnConnectTaskListener listener) {
            mListener = listener;
        }

        public interface OnConnectTaskListener {
            void onSuccess();

            void onFailure();
        }
    }
}

File: app/src/main/java/com/rpinferetti/picar/TCPSocket.kt
package com.rpinferetti.picar

import android.os.AsyncTask
import android.os.Parcel
import android.os.Parcelable
import android.util.Log

import java.io.BufferedWriter
import java.io.IOException
import java.io.OutputStreamWriter
import java.io.Writer
import java.net.InetAddress
import java.net.Socket
import java.nio.charset.StandardCharsets

class TCPSocket : MySocket, Parcelable {

    private var mAddress = InetAddress.getByName(address)
    private var mSocket: Socket? = null
    private var mBufferOut: Writer? = null

    constructor(address: String, port: Int) : super(address, port)

    override fun connect() {
        val task = ConnectTask(mAddress, port, object : ConnectTask.OnConnectTaskListener {
            override fun onSuccess(socket: Socket, bufferOut: Writer?) {
                mSocket = socket
                mBufferOut = bufferOut
                onMySocketListener?.onConnectSuccess()
            }

            override fun onFailure() {
                onMySocketListener?.onConnectFailure()
            }
        })

        task.execute()
    }

    override fun sendMessage(msg: String) {
        if (mSocket != null) {
            Thread(Runnable {
                mBufferOut?.let {
                    try {
                        it.write(msg)
                        it.flush()
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                }
            }).start()
        }
    }

    override fun disconnect() {
        try {
            mBufferOut?.close()
            mSocket?.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }


    // CONNECT ASYNCTASK

    class ConnectTask internal constructor(
            private val address: InetAddress,
            private val port: Int,
            private val listener: OnConnectTaskListener
    ) : AsyncTask<Unit, Unit, Unit>() {
        private var socket: Socket? = null
        private var bufferOut: Writer? = null

        override fun doInBackground(vararg units: Unit): Unit? {
            try {
                socket = Socket(address, port)
                bufferOut = BufferedWriter(
                        OutputStreamWriter(
                                socket!!.getOutputStream(),
                                StandardCharsets.UTF_8
                        )
                )
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return Unit
        }

        override fun onPostExecute(aUnit: Unit) {
            super.onPostExecute(aUnit)
            if (socket != null)
                listener.onSuccess(socket!!, bufferOut)
            else
                listener.onFailure()
        }

        interface OnConnectTaskListener {
            fun onSuccess(socket: Socket, bufferOut: Writer?)

            fun onFailure()
        }
    }

    // PARCELABLE IMPLEMENTATION

    constructor(parcel: Parcel) : super(parcel)

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        super.writeToParcel(dest, flags)
    }

    override fun describeContents(): Int {
        return 0
    }


    companion object CREATOR : Parcelable.Creator<TCPSocket> {
        override fun createFromParcel(parcel: Parcel): TCPSocket {
            return TCPSocket(parcel)
        }

        override fun newArray(size: Int): Array<TCPSocket?> {
            return arrayOfNulls(size)
        }
    }
}
